File: csc/2016/kve/src/main/SynchronizedQueue.java
package main;

import java.util.Iterator;
import java.util.LinkedList;

public class SynchronizedQueue<E> {
    private final LinkedList<E> queue = new LinkedList<>();
    private static final Object monitor = new Object();

    public void put(E item) {
        synchronized (monitor) {
            queue.add(item);
            if (queue.size() == 1) {
                monitor.notify();
            }
        }
    }

    public E pop() throws InterruptedException {
        synchronized (monitor) {
            while (queue.isEmpty()) {
                monitor.wait();
            }

            return queue.pop();
        }
    }

    public Iterator<E> iterator() {
        return queue.listIterator();
    }

    public boolean isEmpty() {
        return queue.isEmpty();
    }
}

File: csc/2016/kve/src/collections/SynchronizedQueue.kt
package collections

import java.util.*

@Suppress("PLATFORM_CLASS_MAPPED_TO_KOTLIN")
class SynchronizedQueue<E> {
    private val queue = LinkedList<E>()

    @Synchronized fun put(item: E) {
        queue.add(item)
        if (queue.size == 1) {
            (this as Object).notify()
        }
    }

    @Synchronized @Throws(InterruptedException::class)
    fun pop(): E {
        while (queue.isEmpty()) {
            (this as Object).wait()
        }

        return queue.pop()
    }
}
